Adds minikube support for pre-built images usage (#1)
accuknox/cilium-spire-tutorials: scenario03/README.md

## Scenario 3: Upgrading non-secure connections to mTLS

> 

Reviewer: ```suggestion

Description of the scenario: the goal of the scenario exposed by the image below is to upgrade the the connection from HTTP to HTTPs. This tutorial uses the Star Wars scenario from Cilium, and, based on it, upgrade the connection between the pods `xwing` to `deathstart`. A Cilium Network Policy (CNP) is going to be applied to upgrade the connection. For this tutorial the following steps will be performed:

1. Deploy the Star Wars scenario;
2. Apply a CNP to upgrade the connection between `xwing` and `deathstart`;
3. Do a HTTP test connection.

Note: for simplicity, in this tutorial, the upgrade happens just for HTTP connection **originated** by `xwing`.
```

<img src="../imgs/scenario03.png" alt="drawing" width="800"/>

Create spire registration entries:

```

> ./0-create_registration_entries.sh

Reviewer: ```suggestion
% ./0-create_registration_entries.sh
```

```

> Deploy starwars pods (from cilium tutorials):

Reviewer: ```suggestion
Deploy Star Wars scenario (from Cilium tutorials):
```

```

> kubectl apply -f 1-http-sw-app.yaml

Reviewer: ```suggestion
% kubectl apply -f 1-http-sw-app.yaml
```

```

> Check if the SPIFFE ID was assigned in `swing` and `deathstar` pods:

Reviewer: ```suggestion
Check if the SPIFFE ID was assigned in `xwing` and `deathstar` pods. The pod `xwing` must have a new label called `spiffe://example.org/xwing` and the pod `deathstar` a new label called `spiffe://example.org/deathstar`.
```

```

> kubectl -n kube-system exec cilium-74m7n -- cilium endpoint list                                                                       

Reviewer: ```suggestion
% kubectl -n kube-system exec cilium-74m7n -- cilium endpoint list                                                                       
```

                                                           k8s:io.kubernetes.pod.namespace=kube-system                                                           
                                                           k8s:k8s-app=kube-dns                                                                                   
3362       Disabled           Disabled          12147      k8s:app=spire-server                                                             10.0.0.140   ready   
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                              
                                                           k8s:io.cilium.k8s.policy.serviceaccount=spire-server                                                  
                                                           k8s:io.kubernetes.pod.namespace=spire                                                                 
                                                           k8s:statefulset.kubernetes.io/pod-name=spire-server-0     
```

> Now, we need to enforce a policy in order to upgrade the connection for the egress traffics from `swing` pod to port 80 and downgrade the connection for the ingress traffic to port 80 in `deathstar`.

Reviewer: ```suggestion
Now, we need to enforce a policy in order to upgrade the connection for the egress traffics from `xwing` pod to port 80 and downgrade the connection for the ingress traffic to port 80 in `deathstar`.
```

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "tls-upgrade-xwing"
spec:
  endpointSelector:
    matchLabels:
      class: xwing
  egress:
  - toPorts:
    - ports:
      - port: "80"
        protocol: "TCP"
      originatingTLS:
        spiffe:
          peerIDs:
            - spiffe://example.org/deathstar
      rules:
        http:
        - {}
---
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "tls-upgrade-deathstar"
spec:
  endpointSelector:
    matchLabels:
      class: deathstar
  ingress:
  - toPorts:
    - ports:
      - port: "80"
        protocol: "TCP"
      terminatingTLS:
        spiffe:
          peerIDs:
            - spiffe://example.org/xwing
      rules:
        http:
        - {}
---
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "enable-xwing-dns"
spec:
  description: "Enable DNS traffic for xwing"
  endpointSelector:
    matchLabels:
     org: alliance
     class: xwing
  egress:
    - toPorts:
      - ports:
        - port: "53"
          protocol: UDP 
```

> The last policy is related to allow the DNS traffic for swing. It will be used later for HTTP request.

> Apply Cilium Network Policies  (CNPs):

Reviewer: ```suggestion
Apply Cilium Network Policies (CNP):
```

```

> kubectl apply -f 2-mtls-upgrade.yaml

Reviewer: ```suggestion
% kubectl apply -f 2-mtls-upgrade.yaml
```

```

> Check if the policies were enforced in `swing` and `deathstar` endpoints:

Reviewer: ```suggestion
Check if the policies were enforced in `xwing` and `deathstar` endpoints. Just to remember, for pod `xwing` we apply a policy just for egress and for pod `deathstar`, just for ingress. Just keep it in mind when looking at the column `ENFORCEMENT`.
```

```

> kubectl exec cilium-74m7n -- cilium endpoint list                                                                       

Reviewer: ```suggestion
% kubectl exec cilium-74m7n -- cilium endpoint list                                                                       
```

                                                           k8s:io.kubernetes.pod.namespace=kube-system                                                           
                                                           k8s:k8s-app=kube-dns                                                                                   
3362       Disabled           Disabled          12147      k8s:app=spire-server                                                             10.0.0.140   ready   
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                              
                                                           k8s:io.cilium.k8s.policy.serviceaccount=spire-server                                                  
                                                           k8s:io.kubernetes.pod.namespace=spire                                                                 
                                                           k8s:statefulset.kubernetes.io/pod-name=spire-server-0  
```

> Send landing request:

Reviewer: ```suggestion
Send landing request. The following script is going to perform a HTTP request. This connection is going to be upgrade to HTTPS. 
```

```
./4-curl.sh
```

> The command `Ship landed` will be shown if the execution was succeeded:

Reviewer: ```suggestion
If the execution was succeeded, the command will return `Ship landed`.

Note: one simple way to verify the encryption traffic is the following. Without the policy applied, the traffic from this lab dont go through cilium_host. After applying the policy, the cilium_host receives the encrypted traffic.
1. Login in minikube `minikube ssh`;
2. Inside minikube, download tcpdump `apt-get update & apt-get install tcpdump`;
3. Capture the traffic from the interface cilium_host `tcpdump -i cilium_host port 80`.

```
